pkg/datapath/loader/base.go

```diff
@@ -0,0 +1,303 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/cgroups"
+	"github.com/cilium/cilium/pkg/command/exec"
+	"github.com/cilium/cilium/pkg/datapath"
+	"github.com/cilium/cilium/pkg/datapath/alignchecker"
+	"github.com/cilium/cilium/pkg/datapath/prefilter"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	initArgLib int = iota
+	initArgRundir
+	initArgIPv4NodeIP
+	initArgIPv6NodeIP
+	initArgMode
+	initArgDevice
+	initArgDevicePreFilter
+	initArgModePreFilter
+	initArgMTU
+	initArgIPSec
+	initArgMasquerade
+	initArgEncryptInterface
+	initArgHostReachableServices
+	initArgHostReachableServicesUDP
+	initArgCgroupRoot
+	initArgBpffsRoot
+	initArgNodePort
+	initArgMax
+)
+
+func (l *Loader) writeNetdevHeader(dir string, o datapath.BaseProgramOwner) error {
+	headerPath := filepath.Join(dir, common.NetdevHeaderFileName)
+	log.WithField(logfields.Path, headerPath).Debug("writing configuration")
+
+	f, err := os.Create(headerPath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %s for writing: %s", headerPath, err)
+
+	}
+	defer f.Close()
+
+	if err := l.templateCache.WriteNetdevConfig(f, o); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Must be called with option.Config.EnablePolicyMU locked.
+func writePreFilterHeader(preFilter *prefilter.PreFilter, dir string) error {
+	headerPath := filepath.Join(dir, common.PreFilterHeaderFileName)
+	log.WithField(logfields.Path, headerPath).Debug("writing configuration")
+	f, err := os.Create(headerPath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %s for writing: %s", headerPath, err)
+
+	}
+	defer f.Close()
+	fw := bufio.NewWriter(f)
+	fmt.Fprint(fw, "/*\n")
+	fmt.Fprintf(fw, " * XDP device: %s\n", option.Config.DevicePreFilter)
+	fmt.Fprintf(fw, " * XDP mode: %s\n", option.Config.ModePreFilter)
+	fmt.Fprint(fw, " */\n\n")
+	preFilter.WriteConfig(fw)
+	return fw.Flush()
+}
+
+// CompileBasePrograms compiles the base programs managed by this loader.
+func (l *Loader) CompileBasePrograms(o datapath.BaseProgramOwner, deviceMTU int, iptMgr datapath.RulesManager, p datapath.Proxy, r datapath.RouteReserver) error {
```

Comment:
Given that there's a bit more than just compiling base programs (I see a rules manager, a proxy and a routeReserver), this seems almost like a:

```suggestion
// Reinitialize (re-)configures the base datapath configuration including global BPF programs, netfilter rule configuration and reserving routes in IPAM for locally detected prefixes. It may be run upon initial Cilium startup, after restore from a previous Cilium run, or during regular Cilium operation.
func (l *Loader) Renitialize(o datapath.BaseProgramOwner, deviceMTU int, iptMgr datapath.RulesManager, p datapath.Proxy, r datapath.RouteReserver) error {
```

(Although I wonder if the IPAM bits are only here because it was a convenient place for it when this was all in daemon; I'm not sure that it's logically actually part of the datapath).

Reply:
Had to rebase, so used your suggestion.